### ISSUE TITLE
File Upload: Batch upload problems: Drag and Drop, and "select files" feature is skipping files, zip not unpacking.

### ISSUE THREAD

**Issue by sbarbosadataverse** [IQSS/dataverse#3116]
The data are yet to be published so I can't paste the example here, but visually you see and are presented a count of how many files are being dragged, and you can see all files listed to begin upload, with blue bars starting to process, and in the end only a selection of the files are actually uploaded. I tested this with drag and drop and with "Select files" to upload. Both are skipping files.

**Comment by kcondon** [comment]
This was reported by Sonia and confirmed by me using only 2 files while this was happening. It appears to be a timing issue that is made more apparent when the server is in a certain state since it generally works.

The symptom is: select more than one file to drag and drop, either on dataset create or edit, and only one of the files or at least not all of the files are uploaded and it is unpredictable which file it is.

Here is the associated server log error:

[2016-05-12T14:57:26.611-0400] [glassfish 4.1] [WARNING] [] [javax.enterprise.resource.webcontainer.jsf.lifecycle] [tid: ThreadID=458 ThreadName=jk-connector(6)] [timeMillis: 1463079446611] [levelValue: 900] [[
/editFilesFragment.xhtml @51,171 fileUploadListener="#{EditDatafilesPage.handleFileUpload}": javax.ejb.EJBException
javax.el.ELException: /editFilesFragment.xhtml @51,171 fileUploadListener="#{EditDatafilesPage.handleFileUpload}": javax.ejb.EJBException
at com.sun.faces.facelets.el.TagMethodExpression.invoke(TagMethodExpression.java:111)
at org.primefaces.component.fileupload.FileUpload.broadcast(FileUpload.java:318)
at javax.faces.component.UIViewRoot.broadcastEvents(UIViewRoot.java:755)
at javax.faces.component.UIViewRoot.processDecodes(UIViewRoot.java:931)
at com.sun.faces.lifecycle.ApplyRequestValuesPhase.execute(ApplyRequestValuesPhase.java:78)
at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:101)
at com.sun.faces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:198)
at javax.faces.webapp.FacesServlet.service(FacesServlet.java:646)
at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1682)
at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:344)
at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
at org.glassfish.tyrus.servlet.TyrusServletFilter.doFilter(TyrusServletFilter.java:295)
at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
at org.ocpsoft.rewrite.servlet.RewriteFilter.doFilter(RewriteFilter.java:205)
at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:256)
at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:316)
at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:160)
at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:734)
at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:673)
at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:174)
at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:734)
at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:673)
at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:412)
at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:282)
at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:459)
at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:167)
at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:201)
at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:175)
at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:235)
at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:561)
at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:565)
at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:545)
at java.lang.Thread.run(Thread.java:745)
Caused by: javax.ejb.EJBException
at com.sun.ejb.containers.EJBContainerTransactionManager.processSystemException(EJBContainerTransactionManager.java:748)
at com.sun.ejb.containers.EJBContainerTransactionManager.completeNewTx(EJBContainerTransactionManager.java:698)
at com.sun.ejb.containers.EJBContainerTransactionManager.postInvokeTx(EJBContainerTransactionManager.java:503)
at com.sun.ejb.containers.BaseContainer.postInvokeTx(BaseContainer.java:4566)
at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2074)
at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2044)
at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:220)
at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:88)
at com.sun.proxy.$Proxy297.createDataFiles(Unknown Source)
at edu.harvard.iq.dataverse.ingest.__EJB31_Generated__IngestServiceBean__Intf____Bean.createDataFiles(Unknown Source)
at edu.harvard.iq.dataverse.EditDatafilesPage.handleFileUpload(EditDatafilesPage.java:1204)
at sun.reflect.GeneratedMethodAccessor1251.invoke(Unknown Source)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at com.sun.el.parser.AstValue.invoke(AstValue.java:289)
at com.sun.el.MethodExpressionImpl.invoke(MethodExpressionImpl.java:304)
at org.jboss.weld.util.el.ForwardingMethodExpression.invoke(ForwardingMethodExpression.java:40)
at org.jboss.weld.el.WeldMethodExpression.invoke(WeldMethodExpression.java:50)
at com.sun.faces.facelets.el.TagMethodExpression.invoke(TagMethodExpression.java:105)
... 45 more
Caused by: java.util.ConcurrentModificationException
at java.util.Vector$Itr.checkForComodification(Vector.java:1184)
at java.util.Vector$Itr.next(Vector.java:1137)
at org.eclipse.persistence.indirection.IndirectList$1.next(IndirectList.java:571)
at edu.harvard.iq.dataverse.ingest.IngestServiceBean.checkForDuplicateFileNames(IngestServiceBean.java:632)
at edu.harvard.iq.dataverse.ingest.IngestServiceBean.createSingleDataFile(IngestServiceBean.java:806)
at edu.harvard.iq.dataverse.ingest.IngestServiceBean.createSingleDataFile(IngestServiceBean.java:790)
at edu.harvard.iq.dataverse.ingest.IngestServiceBean.createDataFiles(IngestServiceBean.java:596)
at sun.reflect.GeneratedMethodAccessor1252.invoke(Unknown Source)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:1081)
at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:1153)
at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4786)
at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:656)
at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:64)
at org.jboss.weld.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
at sun.reflect.GeneratedMethodAccessor113.invoke(Unknown Source)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
at sun.reflect.GeneratedMethodAccessor114.invoke(Unknown Source)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:369)
at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4758)
at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4746)
at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:212)
... 57 more
]]

**Comment by kcondon** [comment]
@sbarbosadataverse Can you take a look at this and update this ticket? Thanks.

**Comment by kcondon** [comment]
OK, tested 2 scenarios, both user reported:
1. Drag and drop is missing files.
2. Uploading zip file does not work.

Tested on both the current version in production, 4.3.1 and the release in progress, 4.4, which contains a fix for ConcurrentModificationException.

This is what we found:
I. Drag and drop missing files:
-both v4.3.1 and v4.4 showed the same failure of missing files when moving two batches of 15 small (43k) pdf files, saving in between each batch. The error in the server log was ConcurrentModificationException.
-thinking it may have to do with thumbnail generation we tried uploading multiple batches without saving in between batches. v4.3.1 showed the same error when uploading the second batch. v4.4 seemed to work better in this case, not showing the failure until the 4th batch of 15.
-Note: another symptom is with each batch uploaded, the file upload complete progress slowed noticeably until it was very slow.

II. Upload zip not working
-we first tested with a user's file that was failing: 1.3GB containing 400, 4.2MB tiff files on v4.4. The symptom was the zip file showed the progress bar, got to the end in a reasonable time but then stayed there and did not appear to complete or unpack.
-we then zipped up the 4, 15 pdf file batches from the first test to see whether smaller zip files might be a good workaround for drag and drop. We tested on v4.4 and found the first 3 batch files unpacked well but the 4th froze as with the larger zip file. We finally just decided to save and found all files had actually been uploaded.
-we then decided to try saving the large zip file that appeared frozen. The screen grayed out as it does during an update but did not seem to complete the save as it stayed on the grayed out screen. We opened the dataset in a separate tab and and all 400 files were saved. So, this seems to indicate the file upload, unpack, and saved states are not always being communicated and the page updating.

**Comment by landreev** [comment]
@kcondon - Kevin, is the large zip file from a user that you mentioned above still available somewhere?

**Comment by kcondon** [comment]
@landreev I do not have the files readily available but I tried to describe their composition in the May 23rd comment: for drag and drop, not zipped: 1. 2 batches of 15, 43k pdf files, 2. the 1000 test files that you made, unpacked I believe, for zip file issue: 3. zip file containing 400 4.2MB tiff files. 

Sonia may know which customer she was doing this batch processing for. The 1000 file zip you have. Hope it helps. When I'm back at work I can check my local for copies.

**Comment by landreev** [comment]
@kcondon, @pdurbin 
I have made a pull request (3459; https://github.com/IQSS/dataverse/pull/3459) and I'm dragging the issue into code review. 
HOWEVER, there is one issue already in QA - #3454, Phil's fix of a small bug introduced in the directorylabel PR. It's a two line fix, but it is in a utility class that my PR modifies heavily. (Plus I want to rename that utility). 

So basically I wanted to  ask to have that PR quickly merged into develop, and then resync my branch, before it is code-reviewed. HOWEVER, now I'm leaning towards requesting that we simply ignore and delete Phil's PR (3457), and go directly to code review and QA my branch instead (I have added a fix for the directorylabel bug on my own). But then I would still want to rename the utility in question... So let's discuss tomorrow.

**Comment by pdurbin** [comment]
@landreev sure, we can discuss tomorrow, but one change at in PR #3457 that's missing from PR #3459 is this change to FileMetadata.java:

```
-    private String directoryLabel = "";
+    private String directoryLabel;
```

I think it's a nice change and that we should incorporate it into PR #3459. I'm not a big fan of persisting empty strings in the database when these values could just be null instead.

In PR #3457 I had also consolidated 3354-alt-checksum.sql and 3405-filemetadata-directory-label.sql into upgrade_v4.5.1_to_v4.6.sql. This is less important to me but should be done at some point before the 4.6 release.

**Comment by landreev** [comment]
@pdurbin 
Yes, thanks for pointing this out! Adding this to my branch...

**Comment by pdurbin** [comment]
@landreev in PR #3457 there's also a fix for the failing "FileMetadataIT" test:

https://github.com/IQSS/dataverse/blob/198891e3c8f3b98c5ba578b23cf5ae722ce70346/src/test/java/edu/harvard/iq/dataverse/api/FileMetadataIT.java#L153-L154

The is the change from BAD_REQUEST to FORBIDDEN we talked about. Here's how the error looks in Jenkins:

https://build.hmdc.harvard.edu:8443/job/phoenix.dataverse.org-apitest-develop/edu.harvard.iq$dataverse/62/testReport/

![screen shot 2016-11-07 at 1 25 04 pm](https://cloud.githubusercontent.com/assets/21006/20070253/ad3befd4-a4ed-11e6-946b-8d7a3f387396.png)

**Comment by pdurbin** [comment]
@landreev I see you incorporated my FileMetadataIT fix into your pull request at 8b2923a . Thanks.

**Comment by djbrooke** [comment]
Removing @raprasad 's assignment here as he left comments in the PR. Thanks Raman!

**Comment by mheppler** [comment]
Related to #2678

**Comment by kcondon** [comment]
I've confirmed the drag and drop skipping files issue is resolved.
Also, no more large zip file hanging on upload.

I am seeing a slightly increasing pause uploading the first file in subsequent drag and drop batches prior to saving all batches at once.

I am seeing a longish save time for the image files I've uploaded -noticeably longer than it took to upload the files themselves. For example 196, 47k pdfs took 2.5mins to save.

Sonia is also testing with same data, according to her own usage patterns and will comment.

**Comment by kcondon** [comment]
From Sonia:

Test 1:
1st batch drag and drop: 36 files-quick--no saving yet
2nd batch drag and drop: 63 files-quick-no saving

3rd batch 120 files drag and drop--slow to process as completing each file, slow to process as uploaded displayed list
I saved here

4th drag and drop-43 no problem
5th drag and drop - 33 no problem
6th drag and drop-188 files very slow to process and display

7th  6 files....very slow to process and display.....saved here at total 271 files

**Comment by kcondon** [comment]
It seems we've made it reliable but performance is still an issue for large amounts of image files.

**Comment by landreev** [comment]
I pulled the ticket back into dev. - this is to quickly address the issues that were found during "real-life" testing with Sonia. The main issues are: 
1) Sonia made a convincing argument that the explicit sort on the list of newly uploaded files (added last week) was actually NOT needed. Plus it looked like adding it may have broken auto-scrolling on the files table. 
2) A more complex issue: Sonia's scenario included uploading large numbers of files that were all graphic images. This brings into play the issue of real-time generation of image thumbnails. This slows the process down noticeably when there are large numbers of these files. 
With PDF images in particular - it turned out that the upload page previews for these had actually been broken for a while (as in, currently in production). Not seeing the previews for PDFs may not be the worst of problems by itself (nobody has complained, actually). But it causes a more serious issue  of the page attempting to re-generate those previews *repeatedly* for all the PDFs on the page. And that causes a noticeable slowdown. 

I may be able to address this in a half of day or so.

**Comment by landreev** [comment]
I'm moving the ticket back into QA. 

Handling of PDFs should be much improved, throughout the process: the processing, the rendering of uploaded files and the final save should be working noticeably faster; 

For non-PDF images, the final save will be much faster. (The processing and rendering stages were already ok, compared to PDFs). 

There shouldn't be any noticeable performance degradation as more files or file batches are uploaded. 

(I am occasionally seeing that pause after the first upload in a new batch... it is definitely intermittent now;  I don't know what it is - some PrimeFaces artefact? It definitely doesn't seem to be as bad as it was before.)

Please keep an eye out for that broken auto-scroller though; and let me know if it reappears.

**Comment by landreev** [comment]
Status update: 

- liveScroll removed; consulted Mike and Gustavo. They agree it's not worth it. It was my own idea to try and use liveScroll on the uploads page; nobody has specifically asked for it. It would be nice to have, if it were working reliably. I did not realize just how buggy and sensitive it was. 
Seeing how many people are complaining about it, there's a chance PrimeFaces will fix it in their next release. Then we can revisit it. But with its current state, we just can't spend that much of our time trying to work around their bugs. 

A user uploading 1000 files will have to wait noticeably longer for all 1000 to get rendered on the page. But it will still work; and it is still better than how it was before. 

Note that there's still a scroller on the list; making the page somewhat easier to use. It's just not a "live", on-demand scroller. 

- Dropbox uploads have been fixed; all the warning messages should be properly refreshing.
